example/babelstream/src/AlpakaStream.cpp

```diff
@@ -0,0 +1,250 @@
+// Copyright (c) 2015-16 Tom Deakin, Simon McIntosh-Smith,
+// University of Bristol HPC
+//
+// For full license terms please see the LICENSE file distributed with this
+// source code
+//
+// Cupla version created by Jeff Young in 2021
+// Ported from cupla to alpaka by Bernhard Manfred Gruber in 2022
+
+#include "AlpakaStream.h"
+
+constexpr auto TBSIZE = 1024;
+constexpr auto DOT_NUM_BLOCKS = 256;
+
+template<typename T>
+AlpakaStream<T>::AlpakaStream(Idx arraySize, Idx deviceIndex)
+    : arraySize(arraySize)
+    , devHost(alpaka::getDevByIdx<DevHost>(0u))
+    , devAcc(alpaka::getDevByIdx<Acc>(deviceIndex))
+    , sums(alpaka::allocBuf<T, Idx>(devHost, DOT_NUM_BLOCKS))
+    , d_a(alpaka::allocBuf<T, Idx>(devAcc, arraySize))
+    , d_b(alpaka::allocBuf<T, Idx>(devAcc, arraySize))
+    , d_c(alpaka::allocBuf<T, Idx>(devAcc, arraySize))
+    , d_sum(alpaka::allocBuf<T, Idx>(devAcc, DOT_NUM_BLOCKS))
+    , queue(devAcc)
+{
+    if(arraySize % TBSIZE != 0)
+        throw std::runtime_error("Array size must be a multiple of " + std::to_string(TBSIZE));
+    std::cout << "Using alpaka device " << alpaka::getName(devAcc) << std::endl;
+}
+
+struct InitKernel
+{
+    template<typename TAcc, typename T>
+    ALPAKA_FN_ACC void operator()(TAcc const& acc, T* a, T* b, T* c, const T initA, const T initB, const T initC) const
+    {
+        auto const [i] = alpaka::getIdx<alpaka::Grid, alpaka::Threads>(acc);
+        a[i] = initA;
+        b[i] = initB;
+        c[i] = initC;
+    }
+};
+
+template<typename T>
+void AlpakaStream<T>::init_arrays(T initA, T initB, T initC)
+{
+    auto const workdiv = WorkDiv{arraySize / TBSIZE, TBSIZE, 1};
+    alpaka::exec<Acc>(
+        queue,
+        workdiv,
+        InitKernel{},
+        alpaka::getPtrNative(d_a),
+        alpaka::getPtrNative(d_b),
+        alpaka::getPtrNative(d_c),
+        initA,
+        initB,
+        initC);
+    alpaka::wait(queue);
+}
+
+template<typename T>
+void AlpakaStream<T>::read_arrays(std::vector<T>& a, std::vector<T>& b, std::vector<T>& c)
+{
+    alpaka::memcpy(queue, alpaka::createView(devHost, a, static_cast<Idx>(a.size())), d_a);
```

Comment:
I was expecting `alpaka::createView(devHost, a)` to work as well, since `a` is a container, but it failed to compile. Should this work?

Reply:
In general, it should work. I think this is a BUG!

Reply:
Hmm, I had a look and it turns out that this is not a bug. `createview(dev, cont)` requires the `cont` to specialize alpaka extents. I guess what we would need is to provide specializations of `alpaka::GetExtents` and maybe also `alpaka::Dim` for standard containers.

Reply:
Which are already implemented! :)

Reply:
With #1847, this works now.